katsdpimageutils/primarybeamcorrection/primary_beam_correction.py

```diff
@@ -0,0 +1,250 @@
+    # Calling the radial_offset function
+    separation_rad = radial_offset(phase_center, image_wcs, row, col)
+    # Calling the cosine_power_pattern function
+    beam_list = cosine_power_pattern(separation_rad, c_freq)
+    return beam_list
+
+
+def mad(data):
+    """
+    Calculating median absolute deviation (MAD).
+    """
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    """
+    Calculating the standard deviation of each frequency plane using MAD. Rejecting pixels more
+    than 5 sigma from the mean until either no more pixels are rejected or a maximum  of 50
+    iterations is reached.
+    Returns: Weights per each frequency plane
+    """
+    diff = 1
+    i = 0
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return float(0.0), float(0.0)
+    med, sd = mad(data)
+    while diff > 0:
+        i += 1
+        if i > 50:
+            return 1/(sd)**2
+        old_sd = sd
+        old_len = len(data)
+        cut = np.abs(data - med) < 5.0 * sd
+        if np.all(~cut):
+            return 1/(sd)**2
+        data = data[cut]
+        med, sd = mad(data)
+        diff = old_len - len(data)
+        if sd == 0.0:
+            return 1/(old_sd)**2
+    return 1/(sd)**2
+
+
+def weighted_average(arr, weights, mask_image):
+    """
+    Computing weighted average of all the frequency planes.
+    """
+    masked_arr = np.ma.MaskedArray(arr, mask=mask_image)
+    wt_average = np.average(masked_arr, weights=weights, axis=0)
+    return wt_average.data
+
+
+def primary_beam_correction(beam_pattern, path):
+    """
+    Correcting for the primary beam effects.
+    """
+    raw_image = read_fits(path)
+    nterm = raw_image.header['NTERM']
+    snr = []
+    pbc_image = []
+    mask_image = []
+    # Getting all the values with attenuated flux of less than 10% of the peak
+    beam_mask = beam_pattern[-1] <= 0.1
+    for i in range(len((beam_pattern))):
+        beam_pattern[i][beam_mask] = np.nan
+        # Getting the rms noise in each image (before primary beam correction)
+        snr.append(standard_deviation(np.ravel(raw_image.data[0, i+nterm, :, :])))
+        # To correct the effect of the beam we divide by the beam.
+        ratio = np.ravel(raw_image.data[0, i + nterm, :, :]) / beam_pattern[i]
+        mask = np.isnan(ratio)
+        mask_image.append(mask)
+        pbc_image.append(ratio)
+    # Primary beam corrected (pbc) image
+    pbc_image = np.array(pbc_image)
+    mask_image = np.array(mask_image)
+    snr = np.array(snr)
+    # Calculating a weighted average of the individual frequency plane images
+    corr_image = weighted_average(pbc_image, snr, mask_image)
+    # Add new axis
+    corr_image = corr_image.reshape(1, 1, raw_image.data.shape[2], raw_image.data.shape[3])
+    return corr_image
+
+
+def _get_value_from_history(keyword, header):
```

Comment:
This function looks like a good candidate for a unit test.

Reply:
Kim offered her help with unit test, I am learning about it also.